crates/gpui/src/platform/linux/x11/client.rs

```diff
@@ -1735,3 +1763,231 @@ fn xdnd_send_status(
+        multiplier: SCROLL_LINES / fp3232_to_f32(data.increment),
+        scroll_value: old_axis_state_with_valid_scroll_value.and_then(|state| state.scroll_value),
+    }
+}
+
+fn reset_all_pointer_device_scroll_positions(
+    pointer_device_states: &mut BTreeMap<xinput::DeviceId, PointerDeviceState>,
+) {
+    pointer_device_states
+        .iter_mut()
+        .for_each(|(_, device_state)| reset_pointer_device_scroll_positions(device_state));
+}
+
+fn reset_pointer_device_scroll_positions(pointer: &mut PointerDeviceState) {
+    pointer.horizontal.scroll_value = None;
+    pointer.vertical.scroll_value = None;
+}
+
+/// Returns the scroll delta for a smooth scrolling motion event, or `None` if no scroll data is present.
+fn get_scroll_delta_and_update_state(
+    pointer: &mut PointerDeviceState,
+    event: &xinput::MotionEvent,
+    modifiers: &Modifiers,
+) -> Option<Point<f32>> {
+    let horizontal_axis_index = pointer
+        .horizontal
+        .valuator_number
+        .and_then(|valuator_number| get_valuator_axis_index(&event.valuator_mask, valuator_number));
+    let vertical_axis_index = pointer
+        .vertical
+        .valuator_number
+        .and_then(|valuator_number| get_valuator_axis_index(&event.valuator_mask, valuator_number));
+
+    let mut delta_x = None;
+    let mut delta_y = None;
+
+    if let Some(horizontal_axis_index) = horizontal_axis_index {
+        if let Some(value) = event.axisvalues.get(horizontal_axis_index) {
+            delta_x = Some(get_axis_scroll_delta_and_update_state(
+                &mut pointer.horizontal,
+                fp3232_to_f32(*value),
+            ));
+        } else {
+            log::error!(
+                "Encountered invalid XInput valuator_mask, scrolling may not work properly."
+            )
+        }
+    }
+
+    if let Some(vertical_axis_index) = vertical_axis_index {
+        if let Some(value) = event.axisvalues.get(vertical_axis_index) {
+            let delta = get_axis_scroll_delta_and_update_state(
+                &mut pointer.vertical,
+                fp3232_to_f32(*value),
+            );
+            if modifiers.shift {
+                delta_x = Some(delta);
+            } else {
+                delta_y = Some(delta);
+            }
+        } else {
+            log::error!(
+                "Encountered invalid XInput valuator_mask, scrolling may not work properly."
+            )
+        }
+    }
+
+    if delta_x.is_some() || delta_y.is_some() {
+        Some(Point::new(delta_x.unwrap_or(0.0), delta_y.unwrap_or(0.0)))
+    } else {
+        None
+    }
+}
+
+fn get_axis_scroll_delta_and_update_state(axis: &mut ScrollAxisState, new_scroll: f32) -> f32 {
+    let delta_scroll = axis.scroll_value.map_or(0.0, |old_scroll| {
+        (old_scroll - new_scroll) * axis.multiplier
+    });
+    axis.scroll_value = Some(new_scroll);
+    delta_scroll
+}
+
+fn get_valuator_axis_index(valuator_mask: &Vec<u32>, valuator_number: u16) -> Option<usize> {
```

Comment:
This and the following 3-4 methods made me wonder whether we shouldn't introduce a `ValuatorMask` type (that might just wrap `Vec<u32>`) and define the methods on that type.

Reply:
I moved these to `event.rs` since that seems to be where event-related utilities go.

Could do the `ValuatorMask` newtype thing (I'm a big fan of newtypes!), but I'm leaning towards not making that change.  Rationale is:

1. The other 2 methods aren't really specific to valuators, these are operations that are general bitmask operations on `Vec<u32>`.  It does seem weird to put them here, is there a better place to put them?

2. Would only be constructed in one spot and not passed around, so there are no type safety benefits